// @web/test-runner-saucelabs/dist/SauceLabsLauncherManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SauceLabsLauncherManager = void 0;
const saucelabs_1 = __importDefault(require("saucelabs"));
const ip_1 = __importDefault(require("ip"));
class SauceLabsLauncherManager {
    constructor(options, connectOptions) {
        this.launchers = new Set();
        this.closeConnection = async () => {
            if (this.connection == null && this.connectionPromise == null) {
                // already closed
                return;
            }
            if (this.connection == null) {
                // wait for connection to finish opening
                await this.connectionPromise;
            }
            if (this.connection != null) {
                this.connection.close();
            }
            this.connection = undefined;
            this.connectionPromise = undefined;
        };
        this.options = options;
        this.connectOptions = connectOptions;
        this.api = new saucelabs_1.default(this.options);
        process.on('SIGINT', this.closeConnection);
        process.on('SIGTERM', this.closeConnection);
        process.on('beforeExit', this.closeConnection);
        process.on('exit', this.closeConnection);
    }
    get webdriverEndpoint() {
        return `${this.api.webdriverEndpoint}wd/hub`;
    }
    async registerLauncher(launcher) {
        this.launchers.add(launcher);
        if (this.connectionPromise != null) {
            await this.connectionPromise;
            return;
        }
        console.log('[Saucelabs] Setting up Sauce Connect proxy...');
        this.connectionPromise = this.api.startSauceConnect(Object.assign(Object.assign({}, this.connectOptions), { noSslBumpDomains: `127.0.0.1,localhost,${ip_1.default.address()}` }));
        this.connection = await this.connectionPromise;
    }
    async deregisterLauncher(launcher) {
        this.launchers.delete(launcher);
        if (this.launchers.size === 0) {
            this.closeConnection();
        }
    }
}
exports.SauceLabsLauncherManager = SauceLabsLauncherManager;
