// date-fns/esm/constants/index.js
/**
 *  Days in 1 week.
 *
 *  @constant
 *  @type {number}
 *  @default
 */
export var daysInWeek = 7;
/**
 *  Maximum allowed time.
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var maxTime = Math.pow(10, 8) * 24 * 60 * 60 * 1000;
/**
 *  Milliseconds in 1 minute
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var millisecondsInMinute = 60000;
/**
 *  Milliseconds in 1 hour
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var millisecondsInHour = 3600000;
/**
 *  Milliseconds in 1 second
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var millisecondsInSecond = 1000;
/**
 * Minimum allowed time.
 *
 * @name minTime
 * @constant
 * @type {number}
 * @default
 */

export var minTime = -maxTime;
/**
 *  Minutes in 1 hour
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var minutesInHour = 60;
/**
 *  Months in 1 quarter
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var monthsInQuarter = 3;
/**
 *  Months in 1 year
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var monthsInYear = 12;
/**
 *  Quarters in 1 year
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var quartersInYear = 4;
/**
 *  Seconds in 1 hour
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var secondsInHour = 3600;
/**
 *  Seconds in 1 minute
 *
 *  @constant
 *  @type {number}
 *  @default
 */

export var secondsInMinute = 60;
